godotopenxrmeta/src/main/cpp/openxr_fb_render_model.cpp

```diff
@@ -0,0 +1,121 @@
+
+#include <godot_cpp/classes/gltf_document.hpp>
+#include <godot_cpp/classes/gltf_state.hpp>
+#include <godot_cpp/classes/open_xr_interface.hpp>
+#include <godot_cpp/classes/xr_server.hpp>
+#include <godot_cpp/variant/utility_functions.hpp>
+
+using namespace godot;
+
+void OpenXRFbRenderModel::set_render_model_path(String p_render_model_path) {
+	render_model_path = p_render_model_path;
+	load_render_model();
+}
+
+String OpenXRFbRenderModel::get_render_model_path() {
+	return render_model_path;
+}
+
+TypedArray<String> OpenXRFbRenderModel::get_valid_paths() {
+	return OpenXRFbRenderModelExtensionWrapper::get_singleton()->get_valid_paths();
+}
+
+bool OpenXRFbRenderModel::has_render_model() {
+	return render_model != nullptr;
+}
+
+void OpenXRFbRenderModel::load_render_model() {
+	if (render_model != nullptr) {
+		render_model->queue_free();
+		render_model = nullptr;
+	}
+
+	if (render_model_path.is_empty()) {
+		return;
+	}
+
+	PackedByteArray render_model_buffer = OpenXRFbRenderModelExtensionWrapper::get_singleton()->get_buffer(render_model_path);
+	if (render_model_buffer.is_empty()) {
+		UtilityFunctions::print_verbose("Failed to load render model buffer from path [", render_model_path, "] in OpenXRFbRenderModel node");
+		return;
+	}
+
+	Ref<GLTFDocument> gltf_document;
+	gltf_document.instantiate();
+	Ref<GLTFState> gltf_state;
+	gltf_state.instantiate();
+
+	Error err = gltf_document->append_from_buffer(render_model_buffer, "", gltf_state);
+	if (err != OK) {
+		UtilityFunctions::print("Failed to instance render model in OpenXRFbRenderModel node");
+		return;
+	}
+
+	render_model = Object::cast_to<Node3D>(gltf_document->generate_scene(gltf_state));
+	if (render_model) {
+		add_child(render_model);
+		emit_signal("render_model_loaded");
+	}
+}
+
+Node3D *OpenXRFbRenderModel::get_render_model() {
+	return render_model;
+}
+
+void OpenXRFbRenderModel::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_POSTINITIALIZE: {
+			Ref<OpenXRInterface> openxr_interface = XRServer::get_singleton()->find_interface("OpenXR");
+			if (openxr_interface.is_valid()) {
+				openxr_interface->connect("session_begun", callable_mp(this, &OpenXRFbRenderModel::load_render_model));
+			}
+		} break;
+	}
+}
+
+void OpenXRFbRenderModel::_bind_methods() {
+	ClassDB::bind_static_method("OpenXRFbRenderModel", D_METHOD("get_valid_paths"), &OpenXRFbRenderModel::get_valid_paths);
+	ClassDB::bind_method(D_METHOD("has_render_model"), &OpenXRFbRenderModel::has_render_model);
+	ClassDB::bind_method(D_METHOD("get_render_model"), &OpenXRFbRenderModel::get_render_model);
+	ClassDB::bind_method(D_METHOD("set_render_model_path", "render_model_path"), &OpenXRFbRenderModel::set_render_model_path);
+	ClassDB::bind_method(D_METHOD("get_render_model_path"), &OpenXRFbRenderModel::get_render_model_path);
+
+	ADD_PROPERTY(PropertyInfo(Variant::STRING, "render_model_path"), "set_render_model_path", "get_render_model_path");
```

Comment:
Is it possible to help the user enter a valid model path here. A similar issue arises with XRNode3D where there's a pose property string, but the list of poses cannot be statically entered (as it comes from the OpenXR Action Map) and so must be queried at runtime.

XRNode3D implements this by adding the `PROPERTY_HINT_ENUM_SUGGESTION` option as:
https://github.com/godotengine/godot/blob/74c32faa78b54863f8f25c538083907c2bf71791/scene/3d/xr_nodes.cpp#L230
This changes the node-editor to show a drop-down, but still lets the developer click a pen icon and type in a custom string should they choose to.

Then it implements the `_validate_property` method and uses it to populate the hint-strings from the `xr_server->get_suggested_pose_names`:
https://github.com/godotengine/godot/blob/74c32faa78b54863f8f25c538083907c2bf71791/scene/3d/xr_nodes.cpp#L240-L259
You should be able to use the `get_valid_paths` method to populate valid paths.

Reply:
Oh, that's a great idea! I actually didn't realize you could do a dropdown as a suggestion.

However, I don't think using `get_valid_paths()` would work - it'll only return results when we have a running OpenXR session, and that'll only be true in the XR editor. But we could certainly provide all the paths in the spec as suggestions!

Reply:
Worse yet, even if we get it to run in the editor it won't help you, because you'll get a list of valid paths on the PC. On Windows you might get lucky if you've got the Oculus Client configured as OpenXR, but you're dead in the water on Mac.

This is why in the action map our meta data ignores extensions, we can configure the action map regardless of the capabilities of the system on which the user is developing their game. What is important is what the capabilities of the headset are the user is running this on.